File: iOS/Crisis Connect/Features/SOS/SosContactNotifier.swift
//
//  SosContactNotifier.swift
//  Crisis Connect
//
//  Stage 3 of the SOS escalation — the iOS port of Android's SosContactNotifier: automatically
//  message the victim's emergency contacts over the internet relay.
//
//  Recipient selection (v1: automatic only): internet-capable contacts ranked by message volume
//  (more = closer) and by how long they've been a contact (older = closer); the combined rank
//  picks the top three. Alerts go out as template-202 messages (an unmistakable SOS text with a
//  maps link); while the SOS runs, movement streams as template-203 live updates that rewrite
//  the SAME bubble in place on both ends. Stopping the SOS sends a plain-text all-clear; one
//  that can't leave the phone (offline stop) is persisted and retried on the next foreground.
//

import Combine
import CoreLocation
import Foundation

final class SosContactNotifier: ObservableObject, @unchecked Sendable {
    static let shared = SosContactNotifier()

    /// Per-recipient delivery state surfaced on the active-SOS screen so the victim can see who is
    /// being alerted and whether each alert actually went out (Android's SosContactNotifier.Recipient).
    enum RecipientStatus: String, Sendable {
        case waiting
        case sending
        case sent
        case failed
    }

    /// Whether the alerted set came from the victim's hand-picked emergency contacts / confirmed
    /// parents (custom) or the automatic "closest contacts" blend.
    enum SelectionSource: Sendable {
        case custom
        case automatic
    }

    struct Recipient: Identifiable, Equatable, Sendable {
        let sessionId: UUID
        let name: String
        var status: RecipientStatus
        var id: UUID { sessionId }
    }

    /// The contacts this SOS is alerting, with live per-contact status. Observed by the active-SOS UI.
    @Published private(set) var recipients: [Recipient] = []
    /// True once recipient selection finished — distinguishes "still choosing" from "no one to notify".
    @Published private(set) var selectionReady = false
    @Published private(set) var selectionSource: SelectionSource = .automatic

    private static let maxRecipients = 3
    private static let initialLocationWait: TimeInterval = 12
    private static let loopTick: TimeInterval = 15
    private static let initialRetryInterval: TimeInterval = 30
    private static let locationRefreshInterval: TimeInterval = 60
    private static let liveUpdateMinInterval: TimeInterval = 60
    private static let liveUpdateMinMoveMeters: CLLocationDistance = 25
    private static let pendingAllClearKey = "sos.contactNotifier.pendingAllClearSessions"

    private struct AlertedRecipient {
        let contact: ContactRecord
        let alertMessageId: String
    }

    private var job: Task<Void, Never>?
    private var alerted: [UUID: AlertedRecipient] = [:]

    private init() {}

    // MARK: - Recipient-status publishing (all @Published writes hop to the main actor)

    @MainActor
    private func resetRecipientState() {
        recipients = []
        selectionReady = false
    }

    @MainActor
    private func publishSelection(_ targets: [ContactRecord], source: SelectionSource) {
        selectionSource = source
        recipients = targets.map { Recipient(sessionId: $0.id, name: $0.name, status: .waiting) }
        selectionReady = true
    }

    @MainActor
    private func publishStatus(_ sessionId: UUID, _ status: RecipientStatus) {
        guard let index = recipients.firstIndex(where: { $0.sessionId == sessionId }) else { return }
        recipients[index].status = status
    }

    func onSosStarted() {
        guard !PlatformRuntime.isRunningTests else { return }
        job?.cancel()
        alerted = [:]
        // A fresh SOS supersedes any all-clear still waiting from the previous session.
        UserDefaults.standard.removeObject(forKey: Self.pendingAllClearKey)

        job = Task { [weak self] in
            guard let self else { return }
            await self.resetRecipientState()
            let selection = await MainActor.run { Self.selectRecipients() }
            let targets = selection.contacts
            await self.publishSelection(targets, source: selection.source)
            guard !targets.isEmpty else {
                NSLog("SosContactNotifier: no internet-capable contacts to notify")
                return
            }
            var pending: [UUID: ContactRecord] = Dictionary(
                targets.map { ($0.id, $0) },
                uniquingKeysWith: { first, _ in first }
            )

            // A fix makes the first alert far more useful; wait briefly, then run with what we have.
            var lastFix = await Self.waitForFix(timeout: Self.initialLocationWait)
            var lastFixFetchedAt = Date()
            var lastSentFix: SOSSignalLocationPayload.GPSPayload?
            var lastInitialAttemptAt = Date.distantPast
            var lastLiveUpdateAt = Date.distantPast

            while !Task.isCancelled {
                let sessionActive = await MainActor.run { SOSBroadcastManager.shared.isSessionActive }
                guard sessionActive else { break }
                let now = Date()

                // Refresh the fix on its own cadence so the GPS isn't hammered every tick.
                if now.timeIntervalSince(lastFixFetchedAt) >= Self.locationRefreshInterval {
                    if let fix = await MainActor.run(body: {
                        SOSSignalLocationCoordinator.shared.currentVictimLocationPayload()
                    }) {
                        lastFix = fix
                    }
                    lastFixFetchedAt = now
                }

                guard InternetChatTransport.shared.isAvailable() else {
                    try? await Task.sleep(nanoseconds: UInt64(Self.loopTick * 1_000_000_000))
                    continue
                }

                // Initial alerts — rebuilt each attempt so a late send still ships a fresh fix.
                if !pending.isEmpty, now.timeIntervalSince(lastInitialAttemptAt) >= Self.initialRetryInterval {
                    lastInitialAttemptAt = now
                    let text = Self.buildSosMessage(location: lastFix)
                    for contact in Array(pending.values) {
                        await self.publishStatus(contact.id, .sending)
                        let messageId = UUID().uuidString
                        let sent = await InternetChatTransport.shared.sendText(
                            messageId: messageId,
                            text: text,
                            contact: contact,
                            priority: "high",
                            templateCode: InternetChatTransport.sosAlertTemplate
                        )
                        if sent {
                            pending.removeValue(forKey: contact.id)
                            self.alerted[contact.id] = AlertedRecipient(
                                contact: contact, alertMessageId: messageId
                            )
                            lastSentFix = lastFix
                            await MainActor.run {
                                SOSChatStore.shared.ensureSession(
                                    id: contact.id, displayName: contact.name, role: .unknown
                                )
                                _ = SOSChatStore.shared.appendLocalMessage(
                                    sessionId: contact.id,
                                    text: text,
                                    status: .sent,
                                    transportMessageId: messageId,
                                    kind: .sosAlert
                                )
                            }
                            await self.publishStatus(contact.id, .sent)
                        } else {
                            await self.publishStatus(contact.id, .failed)
                        }
                    }
                }

                // Live location updates — rewrite the delivered bubbles when the victim has moved.
                if let fix = lastFix,
                   !self.alerted.isEmpty,
                   now.timeIntervalSince(lastLiveUpdateAt) >= Self.liveUpdateMinInterval,
                   Self.hasMovedEnough(from: lastSentFix, to: fix) {
                    lastLiveUpdateAt = now
                    let newText = Self.buildSosMessage(location: fix)
                    var anySent = false
                    for recipient in self.alerted.values {
                        guard let payload = try? JSONSerialization.data(withJSONObject: [
                            "ref": recipient.alertMessageId,
                            "text": newText
                        ]) else { continue }
                        let sent = await InternetChatTransport.shared.sendText(
                            messageId: UUID().uuidString,
                            text: String(decoding: payload, as: UTF8.self),
                            contact: recipient.contact,
                            priority: "normal",
                            templateCode: InternetChatTransport.sosLocationUpdateTemplate
                        )
                        if sent {
                            anySent = true
                            await MainActor.run {
                                _ = SOSChatStore.shared.updateLocalMessageText(
                                    sessionId: recipient.contact.id,
                                    transportMessageId: recipient.alertMessageId,
                                    text: newText
                                )
                            }
                        }
                    }
                    if anySent { lastSentFix = fix }
                }

                try? await Task.sleep(nanoseconds: UInt64(Self.loopTick * 1_000_000_000))
            }
        }
    }

    /// Cancels alerting and sends the plain-text all-clear to everyone who actually received the
    /// alert. Undeliverable all-clears are persisted and retried on the next app foreground —
    /// nobody should keep worrying after a false alarm.
    func onSosStopped() {
        guard !PlatformRuntime.isRunningTests else { return }
        job?.cancel()
        job = nil
        Task { [weak self] in await self?.resetRecipientState() }
        let toNotify = alerted.values.map(\.contact)
        alerted = [:]
        guard !toNotify.isEmpty else { return }
        UserDefaults.standard.set(
            toNotify.map { $0.id.uuidString },
            forKey: Self.pendingAllClearKey
        )
        Task { [weak self] in
            _ = await self?.deliverPendingAllClear()
            // An offline stop leaves the all-clear persisted; hand it to the network-gated
            // background task so contacts get the all-clear even if the app is never reopened.
            SosFollowUpBackgroundManager.shared.scheduleIfNeeded()
        }
    }

    /// True when an offline stop left an all-clear still awaiting delivery — the trigger the
    /// background follow-up task checks before scheduling itself.
    static var hasPendingAllClear: Bool {
        !(UserDefaults.standard.stringArray(forKey: pendingAllClearKey) ?? []).isEmpty
    }

    /// Attempts delivery of the persisted all-clear messages; called right after the stop and on
    /// app foreground. Never sends while a NEW SOS is running — that session owns the threads.
    @discardableResult
    func deliverPendingAllClear() async -> Bool {
        guard var pendingIds = UserDefaults.standard.stringArray(forKey: Self.pendingAllClearKey),
              !pendingIds.isEmpty else {
            return true
        }
        let sessionActive = await MainActor.run { SOSBroadcastManager.shared.isSessionActive }
        if sessionActive { return true }
        guard InternetChatTransport.shared.isAvailable() else { return false }

        let text = NSLocalizedString("SOS_CONTACT_MESSAGE_ENDED", comment: "")
        for idString in pendingIds {
            guard let sessionId = UUID(uuidString: idString),
                  let contact = await MainActor.run(body: {
                      ContactStore.shared.contacts.first { $0.id == sessionId }
                  }),
                  contact.supportsInternet else {
                pendingIds.removeAll { $0 == idString }
                continue
            }
            let messageId = UUID().uuidString
            let sent = await InternetChatTransport.shared.sendText(
                messageId: messageId,
                text: text,
                contact: contact,
                priority: "high"
            )
            if sent {
                await MainActor.run {
                    _ = SOSChatStore.shared.appendLocalMessage(
                        sessionId: contact.id,
                        text: text,
                        status: .sent,
                        transportMessageId: messageId
                    )
                }
                pendingIds.removeAll { $0 == idString }
            }
        }
        if pendingIds.isEmpty {
            UserDefaults.standard.removeObject(forKey: Self.pendingAllClearKey)
            return true
        }
        UserDefaults.standard.set(pendingIds, forKey: Self.pendingAllClearKey)
        return false
    }

    // MARK: - Selection + message building

    /// Hand-picked emergency contacts when configured; otherwise internet-capable contacts ranked
    /// by message volume and by contact age — a brand-new busy chat and a quiet long-standing one
    /// both surface.
    @MainActor
    private static func selectRecipients() -> (source: SelectionSource, contacts: [ContactRecord]) {
        // Child-profile contacts are never SOS recipients — an alert must reach an adult.
        let contacts = ContactStore.shared.contacts
            .filter { $0.supportsInternet && $0.peerIsChild != true && $0.isAuthorityBridge != true }
        guard !contacts.isEmpty else { return (.automatic, []) }

        // On a child device the confirmed parents ARE the emergency contacts. Only when none are
        // usable fall through to the normal selection — in an emergency reaching someone beats
        // reaching no one.
        if ChildProfileManager.shared.isEnabled {
            let parentIds = ChildProfileManager.shared.parents
            let parentRecipients = contacts.filter { parentIds.contains($0.id) }.prefix(maxRecipients)
            if !parentRecipients.isEmpty { return (.custom, Array(parentRecipients)) }
        }

        let chosen = SosEmergencyContactsStore.load()
        if !chosen.isEmpty {
            let custom = contacts.filter { chosen.contains($0.id) }.prefix(maxRecipients).map { $0 }
            if !custom.isEmpty { return (.custom, custom) }
        }
        let counts = Dictionary(
            contacts.map { ($0.id, SOSChatStore.shared.messageCount(for: $0.id)) },
            uniquingKeysWith: { first, _ in first }
        )
        let byVolume = contacts.sorted { (counts[$0.id] ?? 0) > (counts[$1.id] ?? 0) }
        let byAge = contacts.sorted { $0.createdAt < $1.createdAt }
        let volumeRank = Dictionary(
            byVolume.enumerated().map { ($0.element.id, $0.offset) },
            uniquingKeysWith: { first, _ in first }
        )
        let ageRank = Dictionary(
            byAge.enumerated().map { ($0.element.id, $0.offset) },
            uniquingKeysWith: { first, _ in first }
        )
        let ranked = contacts
            .sorted { (volumeRank[$0.id] ?? 0) + (ageRank[$0.id] ?? 0)
                < (volumeRank[$1.id] ?? 0) + (ageRank[$1.id] ?? 0) }
            .prefix(maxRecipients)
            .map { $0 }
        return (.automatic, ranked)
    }

    private static func buildSosMessage(location: SOSSignalLocationPayload.GPSPayload?) -> String {
        if let location {
            let mapsLink = String(
                format: "https://maps.google.com/?q=%.6f,%.6f",
                locale: Locale(identifier: "en_US_POSIX"),
                location.latitude,
                location.longitude
            )
            return String(
                format: NSLocalizedString("SOS_CONTACT_MESSAGE", comment: ""),
                mapsLink
            )
        }
        return NSLocalizedString("SOS_CONTACT_MESSAGE_NO_LOCATION", comment: "")
    }

    private static func hasMovedEnough(
        from: SOSSignalLocationPayload.GPSPayload?,
        to: SOSSignalLocationPayload.GPSPayload
    ) -> Bool {
        // The very first update after an alert that shipped WITHOUT a fix should always go out.
        guard let from else { return true }
        let distance = CLLocation(latitude: from.latitude, longitude: from.longitude)
            .distance(from: CLLocation(latitude: to.latitude, longitude: to.longitude))
        return distance >= liveUpdateMinMoveMeters
    }

    private static func waitForFix(timeout: TimeInterval) async -> SOSSignalLocationPayload.GPSPayload? {
        let deadline = Date().addingTimeInterval(timeout)
        while Date() < deadline {
            if let fix = await MainActor.run(body: {
                SOSSignalLocationCoordinator.shared.currentVictimLocationPayload()
            }) {
                return fix
            }
            try? await Task.sleep(nanoseconds: 1_500_000_000)
        }
        return nil
    }
}

File: Android/app/src/main/java/com/auralis/crisisconnect/service/sos/SosContactNotifier.kt
package com.auralis.crisisconnect.service.sos

import android.content.Context
import android.content.SharedPreferences
import android.location.Location
import android.util.Log
import com.auralis.crisisconnect.R
import com.auralis.crisisconnect.data.Contact
import com.auralis.crisisconnect.data.getContact
import com.auralis.crisisconnect.data.getContacts
import com.auralis.crisisconnect.data.getSessionFirstMessageTimes
import com.auralis.crisisconnect.data.getSessionMessageCounts
import com.auralis.crisisconnect.data.saveLocalMessage
import com.auralis.crisisconnect.data.saveLocalSosMessage
import com.auralis.crisisconnect.data.updateSosAlertText
import com.auralis.crisisconnect.messaging.InternetChatTransport
import com.auralis.crisisconnect.security.ChildProfileManager
import com.auralis.crisisconnect.service.BlePeerIdentityUtils
import com.auralis.crisisconnect.service.GattSOSServerService
import java.util.Locale
import java.util.UUID
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeoutOrNull
import org.json.JSONObject

/**
 * Stage 3 of the SOS escalation: automatically message the victim's emergency contacts.
 *
 * Recipient selection: the user's hand-picked emergency contacts ([SosEmergencyContactsStore])
 * when set; otherwise an automatic blend of "most messaged" and "longest known" — each contact
 * is ranked on both signals and the combined rank picks the top [MAX_RECIPIENTS]. Only
 * internet-capable contacts qualify either way.
 *
 * Alerts go out as a dedicated SOS message type (template 202 on the wire, SOS_ALERT locally) so
 * chats render them as an unmistakable emergency bubble. While the SOS runs, the victim's
 * movement is streamed as template-203 live updates that rewrite the SAME bubble in place on
 * both ends — contacts always see the latest position without the thread filling up. Stopping
 * the SOS sends a plain-text all-clear; one that can't leave the phone (offline stop) is
 * persisted and delivered by [SosReportWorker] once a network appears.
 */
object SosContactNotifier {

    enum class RecipientStatus { Waiting, Sending, Sent, Failed }

    enum class SelectionSource { Custom, Automatic }

    data class Recipient(
        val sessionCode: String,
        val name: String,
        val status: RecipientStatus
    )

    private data class AlertedRecipient(
        val contact: Contact,
        val alertMessageId: String
    )

    private const val TAG = "SosContactNotifier"
    private const val MAX_RECIPIENTS = SosEmergencyContactsStore.MAX_CONTACTS
    private const val PREFS_NAME = "sos_contact_notifier"
    private const val KEY_PENDING_ALL_CLEAR_SESSIONS = "pending_all_clear_sessions"

    private const val INITIAL_LOCATION_WAIT_MS = 12_000L
    private const val LOOP_TICK_MS = 15_000L
    private const val INITIAL_RETRY_INTERVAL_MS = 30_000L
    private const val LOCATION_REFRESH_INTERVAL_MS = 60_000L
    private const val LIVE_UPDATE_MIN_INTERVAL_MS = 60_000L
    private const val LIVE_UPDATE_MIN_MOVE_METERS = 25f

    private val notifierScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private var job: Job? = null
    private var alerted: MutableMap<String, AlertedRecipient> = linkedMapOf()

    private val _recipients = MutableStateFlow<List<Recipient>>(emptyList())
    val recipients: StateFlow<List<Recipient>> = _recipients.asStateFlow()

    /** True once recipient selection finished (distinguishes "loading" from "no one to notify"). */
    private val _selectionReady = MutableStateFlow(false)
    val selectionReady: StateFlow<Boolean> = _selectionReady.asStateFlow()

    private val _selectionSource = MutableStateFlow(SelectionSource.Automatic)
    val selectionSource: StateFlow<SelectionSource> = _selectionSource.asStateFlow()

    fun onSosStarted(
        context: Context,
        locationProvider: suspend () -> BlePeerIdentityUtils.PeerLocationSnapshot?
    ) {
        val appContext = context.applicationContext
        job?.cancel()
        alerted = linkedMapOf()
        _recipients.value = emptyList()
        _selectionReady.value = false
        // A fresh SOS supersedes any all-clear still waiting from the previous session.
        prefs(appContext).edit().remove(KEY_PENDING_ALL_CLEAR_SESSIONS).apply()
        job = notifierScope.launch {
            val targets = runCatching { selectRecipients(appContext) }
                .onFailure { Log.w(TAG, "Failed to select SOS recipients", it) }
                .getOrDefault(emptyList())
            _recipients.value = targets.map {
                Recipient(it.sessionCode, it.name, RecipientStatus.Waiting)
            }
            _selectionReady.value = true
            if (targets.isEmpty()) {
                Log.d(TAG, "No internet-capable contacts to notify")
                return@launch
            }

            val transport = InternetChatTransport(appContext)
            val pending = targets.associateBy { it.sessionCode }.toMutableMap()

            // A fix makes the first alert far more useful; wait briefly, then run with what we have.
            var lastFix = withTimeoutOrNull(INITIAL_LOCATION_WAIT_MS) {
                runCatching { locationProvider() }.getOrNull()
            }
            var lastFixFetchedAt = System.currentTimeMillis()
            var lastSentFix: BlePeerIdentityUtils.PeerLocationSnapshot? = null
            var lastInitialAttemptAt = 0L
            var lastLiveUpdateAt = 0L

            while (isActive && GattSOSServerService.isDeclared.value) {
                val now = System.currentTimeMillis()

                // Refresh the fix on its own cadence so the GPS isn't hammered every tick.
                if (now - lastFixFetchedAt >= LOCATION_REFRESH_INTERVAL_MS) {
                    runCatching { locationProvider() }.getOrNull()?.let { lastFix = it }
                    lastFixFetchedAt = now
                }

                if (!transport.isAvailable()) {
                    delay(LOOP_TICK_MS)
                    continue
                }

                // Initial alerts — rebuilt each attempt so a late send still ships a fresh fix.
                if (pending.isNotEmpty() && now - lastInitialAttemptAt >= INITIAL_RETRY_INTERVAL_MS) {
                    lastInitialAttemptAt = now
                    val text = buildSosMessage(appContext, lastFix)
                    for (contact in pending.values.toList()) {
                        updateStatus(contact.sessionCode, RecipientStatus.Sending)
                        val messageId = UUID.randomUUID().toString()
                        val sent = runCatching {
                            transport.sendText(
                                messageId = messageId,
                                text = text,
                                createdAtMs = System.currentTimeMillis(),
                                contact = contact,
                                priority = InternetChatTransport.PRIORITY_HIGH,
                                templateCode = InternetChatTransport.SOS_ALERT_TEMPLATE
                            )
                        }.getOrDefault(false)
                        if (sent) {
                            runCatching {
                                saveLocalSosMessage(appContext, contact.sessionCode, messageId, text)
                            }.onFailure {
                                Log.w(TAG, "Failed to persist SOS alert into thread", it)
                            }
                            pending.remove(contact.sessionCode)
                            alerted[contact.sessionCode] = AlertedRecipient(contact, messageId)
                            lastSentFix = lastFix
                            updateStatus(contact.sessionCode, RecipientStatus.Sent)
                        } else {
                            updateStatus(contact.sessionCode, RecipientStatus.Failed)
                        }
                    }
                }

                // Live location updates — rewrite the delivered bubbles when the victim has moved.
                val fix = lastFix
                if (
                    alerted.isNotEmpty() &&
                    fix != null &&
                    now - lastLiveUpdateAt >= LIVE_UPDATE_MIN_INTERVAL_MS &&
                    hasMovedEnough(lastSentFix, fix)
                ) {
                    lastLiveUpdateAt = now
                    val newText = buildSosMessage(appContext, fix)
                    var anySent = false
                    alerted.values.forEach { recipient ->
                        val payload = JSONObject()
                            .put("ref", recipient.alertMessageId)
                            .put("text", newText)
                            .toString()
                        val sent = runCatching {
                            transport.sendText(
                                messageId = UUID.randomUUID().toString(),
                                text = payload,
                                createdAtMs = System.currentTimeMillis(),
                                contact = recipient.contact,
                                priority = InternetChatTransport.PRIORITY_NORMAL,
                                templateCode = InternetChatTransport.SOS_LOCATION_UPDATE_TEMPLATE
                            )
                        }.getOrDefault(false)
                        if (sent) {
                            anySent = true
                            runCatching {
                                updateSosAlertText(appContext, recipient.alertMessageId, newText)
                            }
                        }
                    }
                    if (anySent) {
                        lastSentFix = fix
                    }
                }

                delay(LOOP_TICK_MS)
            }
        }
    }

    /**
     * Cancels alerting and sends the plain-text all-clear to everyone who actually received the
     * alert. Undeliverable all-clears (offline stop, transient failure) are persisted and retried
     * by [SosReportWorker] once a network appears — nobody should keep worrying after a false alarm.
     */
    fun onSosStopped(context: Context) {
        val appContext = context.applicationContext
        job?.cancel()
        job = null
        val toNotify = alerted.values.map { it.contact }
        alerted = linkedMapOf()
        _recipients.value = emptyList()
        _selectionReady.value = false
        if (toNotify.isEmpty()) {
            return
        }
        prefs(appContext).edit()
            .putStringSet(
                KEY_PENDING_ALL_CLEAR_SESSIONS,
                toNotify.map { it.sessionCode }.toSet()
            )
            .apply()
        notifierScope.launch {
            if (!deliverPendingAllClear(appContext)) {
                SosReportWorker.enqueue(appContext)
            }
        }
    }

    /**
     * Attempts delivery of the persisted all-clear messages; used both right after the stop and by
     * [SosReportWorker] retries. Returns true when nothing is left pending. Never sends while a
     * NEW SOS is already running — the fresh session owns those conversations again.
     */
    internal suspend fun deliverPendingAllClear(context: Context): Boolean {
        val appContext = context.applicationContext
        val pendingSessions = prefs(appContext)
            .getStringSet(KEY_PENDING_ALL_CLEAR_SESSIONS, null)
            ?.toMutableSet()
            ?: return true
        if (pendingSessions.isEmpty()) {
            clearPendingAllClear(appContext)
            return true
        }
        if (GattSOSServerService.isDeclared.value) {
            return true
        }
        val transport = InternetChatTransport(appContext)
        if (!transport.isAvailable()) {
            return false
        }
        val text = appContext.getString(R.string.sos_contact_message_ended)
        for (sessionCode in pendingSessions.toList()) {
            val contact = runCatching { getContact(appContext, sessionCode) }.getOrNull()
            if (contact == null || !contact.supportsInternet) {
                pendingSessions.remove(sessionCode)
                continue
            }
            val messageId = UUID.randomUUID().toString()
            val sent = runCatching {
                transport.sendText(
                    messageId = messageId,
                    text = text,
                    createdAtMs = System.currentTimeMillis(),
                    contact = contact,
                    priority = InternetChatTransport.PRIORITY_HIGH
                )
            }.getOrDefault(false)
            if (sent) {
                runCatching { saveLocalMessage(appContext, sessionCode, messageId, text) }
                pendingSessions.remove(sessionCode)
            }
        }
        prefs(appContext).edit()
            .putStringSet(KEY_PENDING_ALL_CLEAR_SESSIONS, pendingSessions)
            .apply()
        return if (pendingSessions.isEmpty()) {
            clearPendingAllClear(appContext)
            true
        } else {
            false
        }
    }

    internal fun clearPendingAllClear(context: Context) {
        prefs(context.applicationContext).edit().remove(KEY_PENDING_ALL_CLEAR_SESSIONS).apply()
    }

    /**
     * Hand-picked emergency contacts when configured; otherwise the automatic blend: contacts are
     * ranked separately by message volume (more = closer) and by conversation age (older = closer),
     * and the sum of both ranks decides — so a brand-new chat with many messages and a quiet
     * long-standing one both surface.
     */
    private suspend fun selectRecipients(context: Context): List<Contact> {
        // Child-profile contacts are never SOS recipients — an alert must reach an adult.
        val contacts = getContacts(context).filter { it.supportsInternet && !it.peerIsChild }
        if (contacts.isEmpty()) {
            _selectionSource.value = SelectionSource.Automatic
            return emptyList()
        }

        // On a child device the confirmed parents ARE the emergency contacts. Only when none
        // are usable (no parents approved yet) fall through to the normal selection — in an
        // emergency reaching someone beats reaching no one.
        if (ChildProfileManager.isEnabled(context)) {
            val parentCodes = ChildProfileManager.parentsFlow(context).value
            val parentRecipients = contacts
                .filter { it.sessionCode in parentCodes }
                .take(MAX_RECIPIENTS)
            if (parentRecipients.isNotEmpty()) {
                _selectionSource.value = SelectionSource.Custom
                return parentRecipients
            }
        }

        val chosen = runCatching { SosEmergencyContactsStore.load(context) }.getOrDefault(emptySet())
        if (chosen.isNotEmpty()) {
            val custom = contacts
                .filter { it.sessionCode in chosen }
                .take(MAX_RECIPIENTS)
            if (custom.isNotEmpty()) {
                _selectionSource.value = SelectionSource.Custom
                return custom
            }
        }

        _selectionSource.value = SelectionSource.Automatic
        val counts = getSessionMessageCounts(context)
        val firstTimes = getSessionFirstMessageTimes(context)
        val byVolume = contacts.sortedByDescending { counts[it.sessionCode] ?: 0 }
        val byAge = contacts.sortedBy { firstTimes[it.sessionCode] ?: Long.MAX_VALUE }
        val volumeRank = byVolume.withIndex().associate { (i, c) -> c.sessionCode to i }
        val ageRank = byAge.withIndex().associate { (i, c) -> c.sessionCode to i }
        return contacts
            .sortedBy { (volumeRank.getValue(it.sessionCode)) + (ageRank.getValue(it.sessionCode)) }
            .take(MAX_RECIPIENTS)
    }

    private fun buildSosMessage(
        context: Context,
        location: BlePeerIdentityUtils.PeerLocationSnapshot?
    ): String {
        return if (location != null) {
            val mapsLink = String.format(
                Locale.US,
                "https://maps.google.com/?q=%.6f,%.6f",
                location.latitude,
                location.longitude
            )
            context.getString(R.string.sos_contact_message, mapsLink)
        } else {
            context.getString(R.string.sos_contact_message_no_location)
        }
    }

    private fun hasMovedEnough(
        from: BlePeerIdentityUtils.PeerLocationSnapshot?,
        to: BlePeerIdentityUtils.PeerLocationSnapshot
    ): Boolean {
        // The very first update after an alert that shipped WITHOUT a fix should always go out.
        if (from == null) return true
        val results = FloatArray(1)
        return runCatching {
            Location.distanceBetween(
                from.latitude,
                from.longitude,
                to.latitude,
                to.longitude,
                results
            )
            results[0] >= LIVE_UPDATE_MIN_MOVE_METERS
        }.getOrDefault(false)
    }

    private fun updateStatus(sessionCode: String, status: RecipientStatus) {
        _recipients.value = _recipients.value.map {
            if (it.sessionCode == sessionCode) it.copy(status = status) else it
        }
    }

    private fun prefs(context: Context): SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
}
